docs/virtual_snake/4_next_hop.md

```diff
@@ -5,48 +5,67 @@ nav_order: 4
         - Set the best key to the peer’s public key;
         - Set the best candidate to the port through which the peer is reachable;
+        - Set the best sequence to **0**;
 6. For each of our routing table entries, to look for any transitive paths that may take the packet closer to the target than any of our direct peering knowledge has provided us:
     1. Skip the routing table entry if either of the following conditions are true:
         - The routing table entry has expired;
         - The **Source port** of the routing table entry refers to the local router;
-    2. If the frame is not a bootstrap message, the **Path public key** is equal to the **Destination public key** and the best key is not equal to the **Destination public key**:
-        - Set the best key to the **Path public key**;
+        - The **Watermark** of the entry has a higher public key than the watermark of the packet;
+        - The **Watermark** of the entry has the same public key but a lower sequence number than the watermark of the packet;
+    2. If the frame is not a bootstrap message, the **Entry public key** is equal to the **Destination public key** and the best key is not equal to the **Destination public key**:
+        - Set the best key to the **Entry public key**;
+        - Set the best candidate to the **Source port** from the entry;
+        - Set the best sequence to the **Sequence number** from the entry;
+    3. If the ordering **Destination public key ＜ Entry public key ＜ Best key** is true:
+        - Set the best key to the **Entry public key**;
         - Set the best candidate to the **Source port** from the entry;
-    3. If the ordering **Destination public key ＜ Path public key ＜ Best key** is true:
-        - Set the best key to the **Path public key**;
-        - Set the best candidate to the **Source port** from the entry.
+        - Set the best sequence to the **Sequence number** from the entry;
+7. For each of the node's directly connected peers (third loop):
+    1. If the best key equals the connected peer’s public key, i.e. we have previously found the peer’s key to be the best key, we can further refine the path to use either the faster or lower latency link type to route to that peer:
+        1. If the **Best candidate** has a slower peer connection type (**Multicast > Remote > Bluetooth**) than the connected peer:
+            - Set the best candidate to the connected peer from the entry;
+        2. If the **Best candidate** has the same peer connection type as the connected peer, the same root sequence number and a higher receive order number for it's tree root announcement than the connected peer:
+            - Set the best candidate to the connected peer from the entry;
+
+A **Watermark** should be returned with the **Best candidate** from the next-hop algorithm. There are two cases to consider for what the watermark should be:
```

Comment:
We might want to stress here that the candidate watermark is used to update the packet before forwarding, otherwise it's not obvious what it's for.